aesd-char-driver/aesd-circular-buffer.c

```diff
@@ -0,0 +1,178 @@
+}
+
+/**
+ * @param buffer the buffer to search for corresponding offset.  Any necessary locking must be performed by caller.
+ * @param char_offset the position to search for in the buffer list, describing the zero referenced
+ *      character index if all buffer strings were concatenated end to end
+ * @param entry_offset_byte_rtn is a pointer specifying a location to store the byte of the returned aesd_buffer_entry
+ *      buffptr member corresponding to char_offset.  This value is only set when a matching char_offset is found
+ *      in aesd_buffer.
+ * @return the struct aesd_buffer_entry structure representing the position described by char_offset, or
+ * NULL if this position is not available in the buffer (not enough data is written).
+ */
+struct aesd_buffer_entry *aesd_circular_buffer_find_entry_offset_for_fpos(struct aesd_circular_buffer *buffer,
+            size_t char_offset, size_t *entry_offset_byte_rtn )
+{
+    if(buffer == NULL || entry_offset_byte_rtn == NULL)
+        return NULL;
+
+/* holds start position, calclated based on current out_offs entry */
+    size_t startPos = 0;
+
+/* calculate desired raw start offset based on current out_offs (entry #) */
+    if(buffer->out_offs == 0)
+        startPos = 0;
+    else
+    {
+        for(int i = 0; i < buffer->out_offs; i++)
+            startPos += buffer->entry[i].size;
+        printf("\nstartPos = %ld", startPos);
+    }
+    printf("\nchar_offset = %ld", char_offset);
+
+    size_t rawOffs      = 0;
+/* holds our current BYTE position as we move through buffer */
+    size_t currPos      = 0;
+/* holds the desired byte offset from the 0th byte of the entry that contains it */
+    size_t posDif       = 0;
+    size_t totalBytes   = 0;
+    uint8_t entryCnt    = 0;
+
+/* true raw offset position we are looking for (ref is 0th byte of entry 0) */
+    rawOffs = startPos + char_offset;
+
+/* loop through all entries and sum total bytes */
+    for(int i = 0; i <= AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED-1; i++)
+        totalBytes += buffer->entry[i].size; 
+
+    printf("\nTotal Bytes: %ld", totalBytes);
+
+/* we dont have that many bytes in the buffer, return NULL */
+    if(char_offset >= totalBytes)
+    {
+        printf("\noffset greater than total bytes, NULL\n\n");
+        return NULL;
+    }
+/* some wacky math to wrap the raw offset byte position if we are reading from non-0 entry */
+    else if(char_offset >= totalBytes - startPos)
+    {
+    /* should give us a negative value, how many bytes back from start we want to read */
+        int fromStartPos = char_offset - totalBytes;
+    /* gives us the true raw offset of the desired byte location */
+        rawOffs = fromStartPos + startPos;
+    }
+
+    printf("\nrawOffs = %ld", rawOffs);
+    printf("\nout_offs: %d\n\n", buffer->out_offs);
+
+/* loop until we find the byte offset we want */
+    entryCnt = 0;
+    while(currPos != rawOffs)
```

Comment:
This works but I think you could simplify the code if you just used a loop like this one, and handle the case where you've iterated through all elements and the position requested is longer than the sum of the size of all entries, rather than pre-calculating this.